addons/delivery/models/sale_order.py

```diff
@@ -111,8 +117,10 @@ def _create_delivery_line(self, carrier, price_unit):
             values['name'] += '\n' + 'Free Shipping'
         if self.order_line:
             values['sequence'] = self.order_line[-1].sequence + 1
-        if self.picking_ids:
-            self.picking_ids.write({'carrier_id': carrier.id})
+
+        to_update = self.picking_ids.filtered(lambda p: p.state not in ['cancel', 'done'] and p.carrier_tracking_ref == False)
+        if to_update:
+            to_update.write({'carrier_id': carrier.id})
```

Comment:
```suggestion
        self.picking_ids.filtered(lambda p: p.state not in ['cancel', 'done'] and not p.carrier_tracking_ref).write({'carrier_id': carrier.id})
```

Reply:
I wrote it explicit that it won't get lost in a one liner. I guess I could remove the condition though, but honestly, it is functional and fixes a bug, right? 😄